src/FrameSemantics.cc

```diff
@@ -282,289 +282,537 @@ static Errors resolveModelPoseWithPlacementFrame(
     }
-    outModel.AddVertex(joint->Name(), sdf::FrameType::JOINT).Id();
   }
 
-  // add frame vertices
-  for (uint64_t f = 0; f < _model->FrameCount(); ++f)
+  /// \brief Children frames.
+  std::vector<FrameWrapper> frames;
+  /// \brief Children models and interface models.
+  std::vector<ModelWrapper> models;
+};
+
+/////////////////////////////////////////////////
+/// \brief Add vertices to either Frame attached to or Pose graph. If the child
+/// element is a model, it calls the corresponding build*Graph function.
+/// \tparam ElementT The type of Element. This must be a class that derives from
+/// WrapperBase.
+/// \tparam GraphT Type of Graph. Either PoseRelativeToGraph or
+/// FrameAttachedToGraph.
+/// \param[in,out] _out The graph to which vertices will be added.
+/// \param[in] _items List of Elements such as links, joints, etc for which
+/// vertices will be created in the graph.
+/// \param[in] _parent Parent element of the items in `_items`.
+/// \param[out] _errors Errors encountered while adding vertices.
+template <typename ElementT, typename GraphT>
+void addVerticesToGraph(ScopedGraph<GraphT> &_out,
+                        const std::vector<ElementT> &_items,
+                        const WrapperBase &_parent, Errors &_errors)
+{
+  for (const auto &item : _items)
   {
-    auto frame = _model->FrameByIndex(f);
-    if (outModel.Count(frame->Name()) > 0)
+    if (_out.Count(item.name) > 0)
     {
-      errors.push_back({ErrorCode::DUPLICATE_NAME,
-          "Frame with non-unique name [" + frame->Name() +
-          "] detected in model with name [" + _model->Name() +
-          "]."});
+      _errors.emplace_back(ErrorCode::DUPLICATE_NAME, item.elementType +
+          " with non-unique name [" + item.name + "] detected in " +
+          lowercase(_parent.elementType) + " with name [" +
+          _parent.name + "].");
       continue;
     }
-    outModel.AddVertex(frame->Name(), sdf::FrameType::FRAME).Id();
+    if constexpr (std::is_same_v<ElementT, ModelWrapper>)
+    {
+      if constexpr (std::is_same_v<GraphT, sdf::FrameAttachedToGraph>)
+      {
+        auto nestedErrors = wrapperBuildFrameAttachedToGraph(_out, item, false);
+        _errors.insert(_errors.end(), nestedErrors.begin(), nestedErrors.end());
+      }
+      else
+      {
+        auto nestedErrors = wrapperBuildPoseRelativeToGraph(_out, item, false);
+        _errors.insert(_errors.end(), nestedErrors.begin(), nestedErrors.end());
+      }
+    }
+    else
+    {
+      _out.AddVertex(item.name, item.frameType);
+    }
   }
+}
 
-  // add nested model vertices
-  for (uint64_t m = 0; m < _model->ModelCount(); ++m)
+/////////////////////////////////////////////////
+/// \brief Add edges to the PoseRelativeTo graph. This handles all ElementT
+/// types except Frame, InterfaceFrame. This also does not handle the case where
+/// ElementT is an InterfaceModelWrapper, but ParentT is either Model or World.
```

Comment:
I think the exceptions in this comment no longer apply. It could say, "This handles all ElementT types, including LinkWrapper, JointWrapper, FrameWrapper, ModelWrapper", your choice whether to list them all

Reply:
Ah, that's correct. I removed them all in 8572ccb756f684c2f1cdf059cfed9eb5a75945f0 since this handles all types.